server/admin.go

```diff
@@ -668,21 +668,31 @@ func (s *adminServer) Drain(ctx context.Context, req *serverpb.DrainRequest) (*s
 
 	nowOn, err := s.server.Drain(on)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	nowOnInts := make([]int32, len(nowOn))
+	res := serverpb.DrainResponse{
+		On: make([]int32, len(nowOn)),
+	}
 	for i := range nowOn {
-		nowOnInts[i] = int32(nowOn[i])
+		res.On[i] = int32(nowOn[i])
+	}
+	if err := stream.Send(&res); err != nil {
+		return err
 	}
+
 	if req.Shutdown {
-		s.server.stopper.Quiesce()
-		go func() {
-			time.Sleep(50 * time.Millisecond)
-			s.server.stopper.Stop()
-		}()
+		go s.server.stopper.Stop()
+	}
+
+	ctx := stream.Context()
+
+	select {
+	case <-s.server.stopper.IsStopped():
```

Comment:
Does the stopper close all the listeners? And the connection is still available and can communicate with the client?